arch/arm/core/swap.c

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright (c) 2018 Linaro, Limited
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <kernel.h>
+#include <toolchain.h>
+#include <kernel_structs.h>
+
+#ifdef CONFIG_EXECUTION_BENCHMARKING
+extern void read_timer_start_of_swap(void);
+#endif
+extern const int _k_neg_eagain;
+
+/**
+ *
+ * @brief Initiate a cooperative context switch
+ *
+ * The __swap() routine is invoked by various kernel services to effect
+ * a cooperative context context switch.  Prior to invoking __swap(), the caller
+ * disables interrupts via irq_lock() and the return 'key' is passed as a
+ * parameter to __swap().  The 'key' actually represents the BASEPRI register
+ * prior to disabling interrupts via the BASEPRI mechanism.
+ *
+ * __swap() itself does not do much.
+ *
+ * It simply stores the intlock key (the BASEPRI value) parameter into
+ * current->basepri, and then triggers a service call exception (svc) to setup
+ * the PendSV exception, which does the heavy lifting of context switching.
+
+ * This is the only place we have to save BASEPRI since the other paths to
+ * __pendsv all come from handling an interrupt, which means we know the
+ * interrupts were not locked: in that case the BASEPRI value is 0.
+ *
+ * Given that __swap() is called to effect a cooperative context switch,
+ * only the caller-saved integer registers need to be saved in the thread of the
+ * outgoing thread. This is all performed by the hardware, which stores it in
+ * its exception stack frame, created when handling the svc exception.
+ *
+ * On Cortex-M0/M0+ the intlock key is represented by the PRIMASK register,
+ * as BASEPRI is not available.
+ *
+ * @return may contain a return value setup by a call to
+ * _set_thread_return_value()
+ *
+ * C function prototype:
+ *
+ * unsigned int __swap (unsigned int basepri);
```

Comment:
Could you adjust this, so the prototype documentation matches the C signature, below?

Or, is it actually not needed, any more, since __swap(..) is already a C function, now?

Reply:
Also, looking at it more carefully: prototype suggested an "unsigned" return, but, now function returns signed "int". Any implications?

Reply:
Ah I thought I had changed it to match the right return.  It is 'unsigned int __swap' in the include.  Fixed.